src/librustdoc/clean/mod.rs

```diff
@@ -2356,57 +2276,41 @@ impl Clean<Item> for doctree::ForeignItem<'_> {
             hir::ForeignItemKind::Type => ForeignTypeItem,
         };
 
-        Item {
-            name: Some(self.name.clean(cx)),
-            attrs: self.attrs.clean(cx),
-            source: self.span.clean(cx),
-            def_id: cx.tcx.hir().local_def_id(self.id).to_def_id(),
-            visibility: self.vis.clean(cx),
-            stability: cx.stability(self.id),
-            deprecation: cx.deprecation(self.id).clean(cx),
-            inner,
-        }
+        Item::from_hir_id_and_parts(self.id, Some(self.name), inner, cx)
     }
 }
 
 impl Clean<Item> for doctree::Macro<'_> {
     fn clean(&self, cx: &DocContext<'_>) -> Item {
-        let name = self.name.clean(cx);
-        Item {
-            name: Some(name.clone()),
-            attrs: self.attrs.clean(cx),
-            source: self.span.clean(cx),
-            visibility: Public,
-            stability: cx.stability(self.hid),
-            deprecation: cx.deprecation(self.hid).clean(cx),
-            def_id: self.def_id,
-            inner: MacroItem(Macro {
+        let what_rustc_thinks = Item::from_def_id_and_parts(
+            self.def_id,
+            Some(self.name),
+            MacroItem(Macro {
                 source: format!(
                     "macro_rules! {} {{\n{}}}",
-                    name,
+                    self.name,
                     self.matchers
                         .iter()
                         .map(|span| { format!("    {} => {{ ... }};\n", span.to_src(cx)) })
                         .collect::<String>()
                 ),
                 imported_from: self.imported_from.clean(cx),
             }),
-        }
+            cx,
+        );
+        // need to override attrs explicitly in case this was inlined
```

Comment:
Not sure what this means.

Reply:
Here are the errors if the attributes aren't overriden:

```
---- [rustdoc] rustdoc/inline_local/macro_by_example.rs stdout ----
12: @has check failed
	`XPATH PATTERN` did not match
	    // @has - '//*[@class="docblock"]' 'docs for foo'

---- [rustdoc] rustdoc/intra-link-disambiguators-removed.rs stdout ----
44: @has check failed
	`XPATH PATTERN` did not match
	// @has - '//a[@href="../intra_link_disambiguators_removed/fn.Name.html"]' "fn@Name"
48: @has check failed
	`XPATH PATTERN` did not match
	// @has - '//a[@href="../intra_link_disambiguators_removed/trait.Name.html"]' "trait@Name"
```

The comment is wrong, it doesn't have anything to do with inlining. The issue is that rustc_metadata strips the attributes on macros: https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/test/rustdoc/intra-link-disambiguators-removed.rs#L42-L49

So this needs to get the attributes from the HIR instead.

Reply:
>The issue is that rustc_metadata strips the attributes on macros

Wait, what.
It shouldn't do that.
`fn encode_info_for_macro_def` writes attributes into metadata, so something else is happening here.

Reply:
Hmm, I'm not sure. Rustdoc isn't doing anything special with the attributes:

https://github.com/rust-lang/rust/blob/75042566d1c90d912f22e4db43b6d3af98447986/src/librustdoc/visit_ast.rs#L622-L625

`def` here is `hir::MacroDef`.

Reply:
The issue was that `hir().attrs()` was ignoring attributes on `MacroDef`. https://github.com/rust-lang/rust/pull/77820/commits/b9f769045e2e04cd361e01c3b67851814a2e4d04